### PR TITLE
[mlir][Symbol] Add verification that symbol's parent is a SymbolTable

### PR DESCRIPTION
**Pull request by caojoshua**
Following the discussion in https://discourse.llvm.org/t/symboltable-and-symbol-parent-child-relationship/75446, we should enforce that a symbol's immediate parent is a symbol table.

I changed some tests to pass the verification. In most cases, we can wrap the func with a module, change the func to another op with regions i.e. scf.if, or change the expected error message.

### CODE REVIEW

File: mlir/include/mlir/IR/SymbolInterfaces.td [hunk 1]
```diff
@@ -171,6 +171,11 @@ def Symbol : OpInterface<"SymbolOpInterface"> {
     if (concreteOp.isDeclaration() && concreteOp.isPublic())
       return concreteOp.emitOpError("symbol declaration cannot have public "
              "visibility");
+    auto parent = $_op->getParentOp();
+    if (parent && !parent->hasTrait<OpTrait::SymbolTable>()) {
```
Comment (joker-eph) [inline mlir/include/mlir/IR/SymbolInterfaces.td:175]:
```suggestion
    if (parent && !parent->hasTrait<OpTrait::SymbolTable>() && parent->isRegistered()) {
```

(We don't know if an unregistered operation implements the symbol table or not)

**Review by joker-eph — APPROVED** [review]
LGTM with a nit.

**Comment by caojoshua** [comment]
I removed the nested function in values.py, which was added in https://reviews.llvm.org/D149902. The test just looks for tests `get_name()`, and we don't need nested funcs to test that.

**Review by makslevental — APPROVED** [review]
> I removed the nested function in values.py, which was added in https://reviews.llvm.org/D149902. The test just looks for tests `get_name()`, and we don't need nested funcs to test that.

Makes sense.

**Pull request merged**